rust/arrow/src/array/builder.rs

```diff
@@ -500,6 +500,49 @@ impl<T: ArrowPrimitiveType> PrimitiveBuilder<T> {
         Ok(())
     }
 
+    /// Appends values from a slice of type `T` and a validity byte slice
+    pub fn append_values(
+        &mut self,
+        values: &[T::Native],
+        is_valid: &[u8],
+        offset: usize,
+    ) -> Result<()> {
```

Comment:
Do you expect users of this function to manually construct `values`, `is_valid` and `offset` separately? If not, it might be better to pass in `ArrayDataRef` as argument so these 3 configs will always be consistent.

Reply:
Thanks, this didn't occur to me, because I started with the signature that was proposed in the comment that led to the JIRA being opened. So by the time I added the offset, I was too far in to notice that I'm passing the elements of ArrayData. I'll change this, and it'll make the interface much better, because the most natural usecase of this is to append arrays to each other

Reply:
Thanks, this didn't occur to me, because I started with the signature that was proposed in the comment that led to the JIRA being opened. So by the time I added the offset, I was too far in to notice that I'm passing the elements of ArrayData. I'll change this, and it'll make the interface much better, because the most natural usecase of this is to append arrays to each other

Reply:
I've given it more thought @houqp . I think adding offsets is out of the scope of the function, because the JIRA was initially created to address the need to manually pass primitive data and its validity map (`append_slice()` but with validity).

I'm thinking a simpler function like the below could work:

```rust
    /// Appends values from a slice of type `T` and a validity boolean slice
    pub fn append_values(
        &mut self,
        values: &[T::Native],
        is_valid: &[bool],
    ) -> Result<()> {
        if values.len() != is_valid.len() {
            return Err(ArrowError::InvalidArgumentError(
                "Value and validity lengths must be equal".to_string(),
            ));
        }
        self.bitmap_builder.append_slice(is_valid)?;
        self.values_builder.append_slice(values)
    }
```

Then we can implement an `append_data(&mut self, data: ArrayDataRef) -> Result<()>;` which generalises over all builder types (with booleans and other types that need specialisation handled separately).

The problem with changing `append_values` above to take `ArrayDataRef` is that there's no guarantee that the underlying data type would meet the `T: ArrowNumericArray` constraint.

@sunchao @paddyhoran this originates from an old comment (https://github.com/apache/arrow/pull/2858/files#r228808948) where the suggestion was to use `memcpy` to efficiently achieve the above. If `memcpy` would work, perhaps it should be applied on `append_slice`? Also, are you fine with `is_valid: &[bool]` instead of it being bitpacked? If it's packed, it's less convenient for everyone as the end-user has to create the validity map, and we have to unpack it.

Reply:
I think that makes sense. Using `&[bool]` also prevents users from blindly passing in the validity bitmap from an existing ArrayData into it without being aware of the offsets. CPP implementation also has a method that takes a vector of bool.

Reply:
append_slice uses `memcpy` underneath, so in your implementation values are already copied using memcpy. memcpy for bitmaps is a lot more complicated due to offsets alignments :(

Reply:
I missed that, I only looked as far as `write_bytes`, but I see that the buffer's `write()` gets called, and that uses memcpy.

Yea, we're going to need that for bitmaps. I want to attempt an `append_data(data: ArrayDataRef)` method as I have the need to merge multiple arrays into 1